sympy/sets/sets.py

```diff
@@ -1605,6 +1699,28 @@ def as_relational(self, symbol):
 
         return And(A_rel, B_rel)
 
+    def _kind(self):
+        try:
+            def _measurement(x):
+                if x is UniversalSet or x is S.UniversalSet:
+                    return float('inf')
+                try:
+                    return len(x)
+                except AttributeError:
+                    return int(x._measure)
+
+            def _ele(x):
+                try:
+                    return x.kind
+                except AttributeError:
+                    return SetKind(NumberKind)
+            if _measurement(self.args[0]) < _measurement(self.args[1]):
+                return SetKind()
+            elif _measurement(self.args[0]) >= _measurement(self.args[1]):
+                return _ele(self.args[0])
+        except TypeError:
+            return self.doit().kind
```

Comment:
The `doit` method should definitely not be called as part of checking the `kind` of an object. Checking `kind` should be a guaranteed cheap operation but `doit` can be arbitrarily expensive.